src/Lib.P2P/Cryptography/Key.cs

```diff
@@ -0,0 +1,187 @@
+        /// </returns>
+        public byte[] Sign(byte[] data)
+        {
+            var signer = SignerUtilities.GetSigner(signingAlgorithmName);
+            signer.Init(true, privateKey);
+            signer.BlockUpdate(data, 0, data.Length);
+            return signer.GenerateSignature();
+        }
+
+        /// <summary>
+        ///   Create a public key from the IPFS message.
+        /// </summary>
+        /// <param name="bytes">
+        ///   The IPFS encoded protobuf PublicKey message.
+        /// </param>
+        /// <returns>
+        ///   The public key.
+        /// </returns>
+        public static Key CreatePublicKeyFromIpfs(byte[] bytes)
+        {
+            var key = new Key();
+
+            var ms = new MemoryStream(bytes, false);
+            var ipfsKey = Serializer.Deserialize<PublicKeyMessage>(ms);
+
+            switch (ipfsKey.Type)
+            {
+                case KeyType.RSA:
+                    key.publicKey = PublicKeyFactory.CreateKey(ipfsKey.Data);
+                    key.signingAlgorithmName = RsaSigningAlgorithmName;
+                    break;
+                case KeyType.Ed25519:
+                    key.publicKey = PublicKeyFactory.CreateKey(ipfsKey.Data);
+                    key.signingAlgorithmName = Ed25519SigningAlgorithmName;
+                    break;
+                case KeyType.Secp256k1:
+                    key.publicKey = PublicKeyFactory.CreateKey(ipfsKey.Data);
+                    key.signingAlgorithmName = EcSigningAlgorithmName;
+                    break;
+                default:
+                    throw new InvalidDataException($"Unknown key type of {ipfsKey.Type}.");
+            }
+
+            return key;
+        }
+
+        /// <summary>
+        ///   Create the key from the Bouncy Castle private key.
+        /// </summary>
+        /// <param name="privateKey">
+        ///   The Bouncy Castle private key.
+        /// </param>
+        public static Key CreatePrivateKey(AsymmetricKeyParameter privateKey)
+        {
+            var key = new Key();
+            key.privateKey = privateKey;
+
+            // Get the public key from the private key.
+            if (privateKey is RsaPrivateCrtKeyParameters rsa)
+            {
+                key.publicKey = new RsaKeyParameters(false, rsa.Modulus, rsa.PublicExponent);
+                key.signingAlgorithmName = RsaSigningAlgorithmName;
+            }
+            else if (privateKey is Ed25519PrivateKeyParameters ed)
+            {
+                key.publicKey = ed.GeneratePublicKey();
+                key.signingAlgorithmName = Ed25519SigningAlgorithmName;
+            }
+            else if (privateKey is ECPrivateKeyParameters ec)
+            {
+                var q = ec.Parameters.G.Multiply(ec.D);
+                key.publicKey = new ECPublicKeyParameters(q, ec.Parameters);
+                key.signingAlgorithmName = EcSigningAlgorithmName;
+            }
+
+            if (key.publicKey == null)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)